Bots/DotNet/Skills/CodeFirst/TeamsWaterfallSkillBot/Bots/SkillBot.cs

```diff
@@ -0,0 +1,227 @@
+
+            // Save any state changes that might have occurred during the turn.
+            await _conversationState.SaveChangesAsync(turnContext, false, cancellationToken);
+        }
+
+        internal static Task<MessagingExtensionActionResponse> CreateCardCommand(MessagingExtensionAction action)
+        {
+            if (action == null)
+            {
+                throw new NullReferenceException(nameof(action));
+            }
+
+            // The user has chosen to create a card by choosing the 'Create Card' context menu command.
+            var createCardData = ((JObject)action.Data).ToObject<CardData>();
+
+            var card = new HeroCard
+            {
+                Title = createCardData.Title,
+                Subtitle = createCardData.Subtitle,
+                Text = createCardData.Text,
+            };
+
+            var attachments = new List<MessagingExtensionAttachment>
+            {
+                new MessagingExtensionAttachment
+                {
+                    Content = card,
+                    ContentType = HeroCard.ContentType,
+                    Preview = card.ToAttachment(),
+                }
+            };
+
+            return Task.FromResult(new MessagingExtensionActionResponse
+            {
+                ComposeExtension = new MessagingExtensionResult
+                {
+                    AttachmentLayout = "list",
+                    Type = "result",
+                    Attachments = attachments,
+                },
+            });
+        }
+
+        internal static Task<MessagingExtensionActionResponse> ShareMessageCommand(MessagingExtensionAction action)
+        {
+            // The user has chosen to share a message by choosing the 'Share Message' context menu command.
+            var heroCard = new HeroCard
+            {
+                Title = $"{action.MessagePayload.From?.User?.DisplayName} orignally sent this message:",
+                Text = action.MessagePayload.Body.Content,
+            };
+
+            if (action.MessagePayload.Attachments?.Count > 0)
+            {
+                // This sample does not add the MessagePayload Attachments.  This is left as an
+                // exercise for the user.
+                heroCard.Subtitle = $"({action.MessagePayload.Attachments.Count} Attachments not included)";
+            }
+
+            // This Messaging Extension example allows the user to check a box to include an image with the
+            // shared message.  This demonstrates sending custom parameters along with the message payload.
+            var includeImage = ((JObject)action.Data)["includeImage"]?.ToString();
+            if (!string.IsNullOrEmpty(includeImage) && bool.TrueString == includeImage)
+            {
+                heroCard.Images = new List<CardImage>
+                {
+                    new CardImage { Url = "https://encrypted-tbn0.gstatic.com/images?q=tbn:ANd9GcQtB3AwMUeNoq4gUBGe6Ocj8kyh3bXa9ZbV7u1fVKQoyKFHdkqU" },
```

Comment:
Let's avoid external dependencies, move the image to the project so the tests are self contained.

Reply:
resolved.